agent/xds/endpoints.go

```diff
@@ -165,52 +165,67 @@ func (s *Server) endpointsFromSnapshotMeshGateway(cfgSnap *proxycfg.ConfigSnapsh
 		resources = append(resources, la)
 	}
 
-	// generate the endpoints for the local service groups
+	// Generate the endpoints for each service and its subsets
 	for svc, endpoints := range cfgSnap.MeshGateway.ServiceGroups {
+		defaultSubsetEndpoints := endpoints
+		defaultSubsetOnlyPassing := false
 		clusterName := connect.ServiceSNI(svc.ID, "", svc.NamespaceOrDefault(), cfgSnap.Datacenter, cfgSnap.Roots.TrustDomain)
-		la := makeLoadAssignment(
-			clusterName,
-			[]loadAssignmentEndpointGroup{
-				{Endpoints: endpoints},
-			},
-			cfgSnap.Datacenter,
-		)
-		resources = append(resources, la)
-	}
-
-	// generate the endpoints for the service subsets
-	for svc, resolver := range cfgSnap.MeshGateway.ServiceResolvers {
-		for subsetName, subset := range resolver.Subsets {
-			clusterName := connect.ServiceSNI(svc.ID, subsetName, svc.NamespaceOrDefault(), cfgSnap.Datacenter, cfgSnap.Roots.TrustDomain)
 
-			endpoints := cfgSnap.MeshGateway.ServiceGroups[svc]
+		// Generate endpoints for each subset cluster. We do this before generating the endpoints for the default/unnamed
+		// subset so that we can take into account the DefaultSubset on the service resolver which may prevent the default/unnamed
+		// cluster from creating endpoints for all service instances.
+		if resolver, hasResolver := cfgSnap.MeshGateway.ServiceResolvers[svc]; hasResolver {
+			for subsetName, subset := range resolver.Subsets {
```

Comment:
random thought while reviewing: I struggled to parse the logic of `endpointsFromSnapshotMeshGateway` for a bit until it all fell in place. In terms of readability, I considered whether it would worthwhile to roll up everything inside of the `if ...; hasResolver` clause into another function, i.e. `subsetEndpointsFromSnapshotMeshGateway`. That gets messy because we then have to deal with the `defaultSubset{Endpoints,OnlyPassing}` variables somehow.

In the end, I decided the improvement to readability might end up being counteracted by messy shenanigans to get the defaultSubset stuff passed through correctly. 🤷‍♂

Reply:
Yeah, the original version was much easier to read although incorrect. I had tried a few different ways of writing this function (not separating it though) and this seemed like the best one at the time. I will see if I can refactor it in such a way to make it easier to read.